### ISSUE TITLE
inner .par collection breaks outer ForkJoinTaskSupport on Linux x86 Scala 2.12.5

### ISSUE THREAD

**Issue by ktegan** [scala/bug#11036]
I first submitted this to stackoverflow and Volodymyr Glushak was not able to recreate the bug on Mac OS or using scastie.  I'm hoping that someone else can recreate this issue or figure out what about my current environment might be causing the problem.

I'm using Scala 2.12.5 and Java OpenJDK 1.8.0_161-b14 on a Linux 3.10.0 x86_64 kernel.

I want to make a parallel collection that uses a fixed number of threads.  The standard advice for this is to set tasksupport for the parallel collection to use a ForkJoinTaskSupport with a ForkJoinPool with a fixed number of threads.  That works fine UNTIL the processing you are doing in your parallel collection itself uses a parallel collection.  When this is the case it appears that the limit set for the ForkJoinPool goes away.

A simple test looks something like the following:

    import java.util.concurrent.atomic.AtomicInteger
    import java.util.concurrent.ForkJoinPool
    import scala.collection.parallel.ForkJoinTaskSupport

    object InnerPar {
      val numTasks = 100
      val numThreads = 10

      /** return maximum number of simultaneous threads running when requesting numThreads */
      def forkJoinPoolMaxThreads(useInnerPar:Boolean): Int = {

        // every thread in the outer collection will increment
        // and decrement this counter as it starts and exits
        val threadCounter = new AtomicInteger(0)

        // this function increments and decrements threadCounter
        // on start and exit, optionally creates an inner parallel collection
        // and finally returns the thread count it found at startup
        def incrementAndCountThreads(idx:Int):Int = {
          val otherThreadsRunning:Int = threadCounter.getAndAdd(1)
          if (useInnerPar) {
            (0 until 20).toSeq.par.map { elem => elem + 1 }
          }
          Thread.sleep(10)
          threadCounter.getAndAdd(-1)
          otherThreadsRunning + 1
        }

        // create parallel collection using a ForkJoinPool with numThreads
        val parCollection = (0 until numTasks).toVector.par
        parCollection.tasksupport = new ForkJoinTaskSupport(new ForkJoinPool(numThreads))
        val threadCountLogList = parCollection.map { idx =>
          incrementAndCountThreads(idx)
        }

        // return the maximum number of simultaneous threads
        // running simultaneously (as counted on each thread start)
        threadCountLogList.max
      } 

      def main(args:Array[String]):Unit = {
        val testConfigs = Seq(true, false, true, false)
        testConfigs.foreach { useInnerPar =>
          val maxThreads = forkJoinPoolMaxThreads(useInnerPar)

          // the total number of threads running should not have exceeded
          // numThreads at any point
          val isSuccess = (maxThreads <= numThreads)

          println(f"useInnerPar $useInnerPar%6s, success is $isSuccess%6s   (maxThreads $maxThreads%4d)") 
        }
      }
    }

And from this in my environment I get the following output, showing that more than numThreads (in the example 10) threads are running simultaneously if we create a parallel collection inside of incrementAndCountThreads().

    useInnerPar   true, success is  false   (maxThreads   22)
    useInnerPar  false, success is   true   (maxThreads   10)
    useInnerPar   true, success is  false   (maxThreads   24)
    useInnerPar  false, success is   true   (maxThreads   10)

Also note that using a ForkJoinTaskSupport in the inner collection does not fix the problem.  In other words you get the same results if you use the following code for the inner collection:

      if (useInnerPar) {
        val innerParCollection = (0 until 20).toVector.par
        innerParCollection.tasksupport = new ForkJoinTaskSupport(new ForkJoinPool(3))
        innerParCollection.map { elem => elem + 1 }
      }

Am I missing something?  If not is there a way to work around this?

Thanks!

**Comment by dlaflamme** [comment]
@SethTisue, what exactly does it mean for this issue to be moved to the Backlog? It seems to me that the behavior observed here completely breaks the user's control of concurrency in their programs and affects the latest two released versions of scala running on the latest version of OpenJDK's 1.8 series.

More information:
It looks like this this problem was introduced between scala 2.12.4 and 2.12.5. I am unable to get ktegan's test to fail on 2.12.4 but it fails consistently on 2.12.5 and 2.12.6 (all tests on the same host/OpenJDK version)

**Comment by SethTisue** [comment]
> what exactly does it mean for this issue to be moved to the Backlog?

“Backlog” issues are issues we won't hold up any particular release over. but any such classification is only provisional

I'd assumed this was a longstanding issue, but if it in fact regressed after 2.12.4, then ah, that's different. so I've moved it to the 2.12.7 milestone (but note that doesn't promise a fix, it just means the issue will at least be considered during release planning)

is there a post-2.12.4 change that's the likely culprit?

**Comment by ktegan** [comment]
Both of these commits are related to ForkJoinPool and are around the right time:

https://github.com/scala/scala/pull/6159/commits/59e104101acdf6a75173b9afa536627bee5f951b
https://github.com/scala/scala/pull/6171/commits/14caff65eefce78105af8fc3e87ba855c77026a6

**Comment by viktorklang** [comment]
@SethTisue I'm not sure how it is intended to work. I don't see how my commit would impact this but please let me know if it seems to be causing this problem.

**Comment by som-snytt** [comment]
I'll take a look.

**Comment by ktegan** [comment]
Hold on a second.  The problem persists whether or not you set the task support for the inner collection (as mentioned with an alternative code snippet at the bottom of the issue).

Also I don't understand your answer.  If the outer tasks are running in a custom pool, then why does creating or not creating an inner parallel collection mutate the behavior of the outer collection?  An immutable parallel collection in Scala should have the same rough concurrency behavior regardless of what code is executing inside a call to map().  This contract is broken (the outer collection appears to be incorrectly mutated) if an inner collection is created inside of one of the threads because all of a sudden the outer parallel collection spawns a many more threads than requested.  The atomicInteger in the test is only measuring the number of threads running in parallel for the outer pool, and so the numbers from the atomicInteger shouldn't really depend on the code executing inside the parallel collection, but the results show otherwise.

**Comment by som-snytt** [comment]
Yes, it's a little mind-bending. The idea is that the outer pool will allocate more threads when it thinks a thread is blocked, i.e., active but not running. When the inner task is run on the same pool, the task is forked, otherwise it submits the task to the other pool. I suspect in the case of submitting to the second pool, joining that task is blocking. Note that it spawns threads to keep `parallelism` number of threads running, not just active (created but not dead).

That's why the counter in the test isn't really counting threads, it's counting tasks. As an implementation detail, those happen to correspond to threads; but it's not counting running threads; imagine those tasks used managed blocking because they're doing I/O, that would explicitly tell the pool to go ahead and grow to keep making progress.

Here's my version of the sample code with println added during the Red Sox game. Comment out setting tasksupport on the "inner" collection to see the bad behavior.

```
package nestedpar

import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.ForkJoinPool
import scala.collection.parallel.ForkJoinTaskSupport

object InnerPar {
  val numTasks = 10
  val numThreads = 4

  /** return maximum number of simultaneous threads running when requesting numThreads */
  def forkJoinPoolMaxThreads(useInnerPar: Boolean): Int = {
    val pool = new ForkJoinPool(
      numThreads,
      /*
      ForkJoinPool.defaultForkJoinWorkerThreadFactory,
      null,
      true,
      4,
      4,
      1,
      null,
      60,
      java.util.concurrent.TimeUnit.SECONDS
      */
    )
    /* Java 9 API
    ForkJoinPool(int parallelism, ForkJoinPool.ForkJoinWorkerThreadFactory factory, Thread.UncaughtExceptionHandler handler, boolean asyncMode, int corePoolSize, int maximumPoolSize, int minimumRunnable, Predicate<? super ForkJoinPool> saturate, long keepAliveTime, TimeUnit unit)
    */
    val taskSupport = new ForkJoinTaskSupport(pool)

    // every thread in the outer collection will increment
    // and decrement this counter as it starts and exits
    val threadCounter = new AtomicInteger(0)

    // this function increments and decrements threadCounter
    // on start and exit, optionally creates an inner parallel collection
    // and finally returns the thread count it found at startup
    def incrementAndCountThreads(idx: Int): Int = {
      println(s"Outer op on: ${Thread.currentThread}")
      val threadsRunning: Int = threadCounter.incrementAndGet()
      if (useInnerPar) {
        val c = (0 until 20).toSeq.par
        c.tasksupport = taskSupport
        println(s"Inner op on: ${Thread.currentThread}, c.tasksupport is ${c.tasksupport}, env is ${c.tasksupport.environment}")
        c.map { elem =>
          println(s"Inner map on: ${Thread.currentThread}")
          elem + 1
        }
        //(0 until 20).toSeq.par.map { elem => elem + 1 }
      } else {
        /* Any execute on a different pool shows the symptom
        List(1).par.map { elem =>
          println(s"Inner map on: ${Thread.currentThread}")
          elem + 1
        }
        */
      }
      Thread.sleep(10L)
      threadCounter.decrementAndGet()
      println(s"Running: $threadsRunning")
      threadsRunning
    }
    // create parallel collection using a ForkJoinPool with numThreads
    val parCollection = (1 to numTasks).toVector.par
    //parCollection.tasksupport = new ForkJoinTaskSupport(new ForkJoinPool(numThreads))
    parCollection.tasksupport = taskSupport
    println(s"parC.tasksupport is ${parCollection.tasksupport}, env is ${parCollection.tasksupport.environment}")
    val threadCountLogList = parCollection.map { idx =>
      incrementAndCountThreads(idx)
    }
    // return the maximum number of simultaneous threads
    // running simultaneously (as counted on each thread start)
    threadCountLogList.max
  }

  def main(args:Array[String]):Unit = {
    println(s"global is ${concurrent.ExecutionContext.global}")
    //val testConfigs = Seq(true, false, true, false)
    val testConfigs = Seq(true)
    testConfigs.foreach { useInnerPar =>
      val maxThreads = forkJoinPoolMaxThreads(useInnerPar)

      // the total number of threads running should not have exceeded
      // numThreads at any point
      val isSuccess = (maxThreads <= numThreads)

      println(f"useInnerPar $useInnerPar%6s, success is $isSuccess%6s   (maxThreads $maxThreads%4d)")
    }
  }
}
```

**Comment by ktegan** [comment]
Thanks, I think I understand things a little bit better now.

So this behavior of "the outer pool will allocate more threads when it thinks a thread is blocked" was added after Scala 2.12.4?  Is there some way to cap the number of simultaneous tasks?  Or alternatively cap the number of active (not running) threads?  In my case each task is allocating part of a finite resource (such as memory), so I really want to cap is tasks not running threads.

**Comment by som-snytt** [comment]
Sorry, I also just noticed your comment about setting tasksupport. It has to be the same pool for the inner collection. It sounds like you just want the old behavior, so I would just do that.

The issue that changed the behavior was where the inner collection has an explicitly set tasksupport, but previously that was ignored.

Otherwise you want the Java 9 FJP as in my comment.

NB: I'll submit a PR to use tasksupport only if it is explicitly set by user. In case that is the preferred ergonomics.

**Comment by viktorklang** [comment]
@som-snytt I also just noticed that there's a possible issue from 2.13.0-M5 and onwards with the following line: https://github.com/scala/scala-parallel-collections/blob/master/core/src/main/scala/scala/collection/parallel/Tasks.scala#L394

Since https://github.com/scala/scala/blob/2.13.x/src/library/scala/concurrent/impl/ExecutionContextImpl.scala#L110

So I venture to say that [Tasks](https://github.com/scala/scala-parallel-collections/blob/master/core/src/main/scala/scala/collection/parallel/Tasks.scala#L394) should look at if the ExecutionContext isInstanceOf ForkJoinPool as well?

**Comment by ktegan** [comment]
Thanks for the comments and help everyone, they are well appreciated.

1. @som-snytt You said that "The issue that changed the behavior was where the inner collection has an explicitly set tasksupport, but previously that was ignored."  However, based on comments by @dlaflamme it sounds like the test shows different behavior between versions 2.12.4 and 2.12.5 when the inner collection does not explicitly set a tasksupport.  Any idea what might account for this change?

2. In the above test if useInnerPar is false the thread still calls Thread.sleep(10) which would seem to block the thread.  However in this case the outer collection does not create more tasks than originally specified.  It is only when an inner parallel collection is created and executed that the behavior changes.  I tried sleeping for 100ms instead and still no change in behvior.  Is there some kind of task (reading from a file?) that I can do to make sure that this issue of the outer collection spawning more tasks than expected is related to some kind of thread blocking operation and isn't specific to using an inner parallel collection?

3. Do I understand correctly that in Java 8 there is no easy way to put a hard cap on the total number of tasks simultaneous tasks, whereas in Java 9 there are extra controls that make this possible?  If so what is the new Java 9 parameter / feature?

4. If anyone has an idea of why you can only create and set tasksupports for about ~1200 collections (https://github.com/scala/bug/issues/11097) any tips would be much appreciated.  A fix would be ideal but right now I'm just not even sure what resource is running out (Java says it cannot create any more native threads).

**Comment by viktorklang** [comment]
@ktegan 3: https://docs.oracle.com/javase/9/docs/api/java/util/concurrent/ForkJoinPool.html#ForkJoinPool-int-java.util.concurrent.ForkJoinPool.ForkJoinWorkerThreadFactory-java.lang.Thread.UncaughtExceptionHandler-boolean-int-int-int-java.util.function.Predicate-long-java.util.concurrent.TimeUnit-

**Comment by ktegan** [comment]
@viktorklang , for (3) it sounds like we're saying that the new maximumPoolSize parameter in Java9 will limit the number of live threads instead of limiting the number of currently running threads as (I'm gathering) the parallelism parameter does.  Thanks!

**Comment by viktorklang** [comment]
@ktegan To put it differently: `parallelism` is a *target*, not a limit.

**Comment by som-snytt** [comment]
@viktorklang Thanks. The architecture makes me feel like I need a "conspiracy wall".

@ktegan The changed behavior is on 2.12.5. The commit page (your link above) shows the tags and also the PR link.

**Comment by ktegan** [comment]
@som-snytt I see the link, but my question is how does the code in the link apply to the current issue.  You said "The issue that changed the behavior was where the inner collection has an explicitly set tasksupport ..."  Based on that I thought you were saying that if the inner collection DOES NOT explicitly set its tasksupport then the PR in the link should have no effect.  If you agree with that statement, then my question is if your PR only affects cases where the inner collection explicitly sets the task support, why are we seeing behavior changes when the inner collection does not explicitly set the the task support?

**Comment by som-snytt** [comment]
@ktegan The original issue was https://github.com/scala/bug/issues/10577 as linked from the PR. It's the same as your example, except the bug was that the tasksupport was ignored. If tasksupport is not set explicitly, you get a default tasksupport.

The documentation is unambiguous:

> Each parallel collection is parametrized with a task support object which is responsible for scheduling and load-balancing tasks to processors.

https://docs.scala-lang.org/overviews/parallel-collections/configuration.html

**Comment by ktegan** [comment]
Sorry, my question isn't coming across well.  Would you expect the PR you did in #10577 to affect cases where the inner collection is not setting the tasksupport?

**Comment by som-snytt** [comment]
Yes, the tasksupport is used whether it was set by user or not. That's why I point out that "each par collection is parameterized with a task support"; I don't see anything which says it is pre-empted in some cases.

My Nota Bene above is about the case where it is not explicitly set.

**Comment by ktegan** [comment]
OK, it seems like there is no simple short term solution to this.  Until Java 9 comes out we might move back to 2.12.4 so that the target thread count is respected 99% of the time instead of 0% of the time.

Thanks for the help!

**Comment by som-snytt** [comment]
@ktegan I like a happy customer. My previous advice to set the same testsupport on both collections is a simple short term solution. That will be 2.12.4 behavior with respect to this issue.

**Comment by ktegan** [comment]
But if the body of the map calls functions that call other functions that call other functions I have to pass the instance of the outer collection's testsupport through all of those.  Basically any function that could be called inside of an parallel collection now has to take an additional testsupport parameter to be able to have the old behavior.

**Comment by som-snytt** [comment]
That is the issue at hand, @ktegan .

**Comment by ktegan** [comment]
@som-snytt  I'm not positive I understand your comment, but I will assume for now you agree with my assessment.  Please let me know if I'm mistaken.

**Comment by SethTisue** [comment]
@ktegan thanks for the report and investigation. this could be invaluable to anyone else who runs into this. thx @som-snytt and @viktorklang as well

**Comment by som-snytt** [comment]
@ktegan sorry I wasn't clear. @SethTisue is laughing at not understanding my comment, nobody understands me. I'm sure there is a classic song along those lines. I was going to say that `Future` passes its context implicitly, but I don't know why parallel collections doesn't, this is not my area of expertise, if I had one. If I can submit a PR, probably it will take a system property to determine which behavior is desired?
